unit/testing-utils/get_goto_model_from_c.cpp

```diff
@@ -0,0 +1,105 @@
+/*******************************************************************\
+
+Module: Get goto model
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+#include "get_goto_model_from_c.h"
+
+#include <ansi-c/ansi_c_language.h>
+
+#include <cbmc/cbmc_parse_options.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+#include <langapi/language_file.h>
+#include <langapi/mode.h>
+
+#include <util/cmdline.h>
+#include <util/config.h>
+#include <util/invariant.h>
+#include <util/message.h>
+#include <util/options.h>
+#include <util/symbol_table.h>
+
+#include <testing-utils/message.h>
+
+namespace
+{
+bool initialized = false;
+optionst options;
+
+void initialize()
+{
+  PRECONDITION(!initialized);
+
+  register_language(new_ansi_c_language);
+
+  cmdlinet cmdline;
+
+  config.main = std::string("main");
+  config.set(cmdline);
+
+  optionst options;
+  cbmc_parse_optionst::set_default_options(options);
+}
+} // namespace
+
+optionalt<goto_modelt> get_goto_model_from_c(std::istream &in)
+{
+  if(!initialized)
+  {
+    initialize();
+    initialized = true;
+  }
+
+  language_filest language_files;
+  language_files.set_message_handler(null_message_handler);
+
+  language_filet &language_file = language_files.add_file("");
+
+  language_file.language = get_default_language();
+
+  languaget &language = *language_file.language;
+  language.set_message_handler(null_message_handler);
+  language.set_language_options(options);
+
+  {
+    const bool error = language.parse(in, "");
+
+    if(error)
+      return {};
```

Comment:
I thinki it’d be better to just throw an exception rather than return optional, because presumably you’d expect for unit tests that their input programs do compile (ditto for all the other error cases)

Reply:
Done